compiler/rustc_expand/src/mbe/macro_parser.rs

```diff
@@ -257,10 +258,10 @@ struct MatcherPos {
     /// against the relevant metavar by the black box parser. An element will be a `MatchedSeq` if
     /// the corresponding metavar decl is within a sequence.
     ///
-    /// It is critical to performance that this is an `Lrc`, because it gets cloned frequently when
+    /// It is critical to performance that this is an `Rc`, because it gets cloned frequently when
     /// processing sequences. Mostly for sequence-ending possibilities that must be tried but end
     /// up failing.
-    matches: Lrc<Vec<NamedMatch>>,
+    matches: Rc<Vec<NamedMatch>>,
```

Comment:
could this be `Rc<[NamedMatch]>`? would it be better?

Reply:
Maybe. I've tried to only change rc stuff. This thing noted as `hot`, so changing vec<->slice can potentially shadow other changes.

Reply:
Okay, we can leave those nits for the future.

Reply:
I just tried it, there are tons of problems.
```text
error[E0277]: the trait bound `[NamedMatch]: Clone` is not satisfied
    --> compiler/rustc_expand/src/mbe/macro_parser.rs:275:37
     |
275  |         let matches = Lrc::make_mut(&mut self.matches);
     |                       ------------- ^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `[NamedMatch]`
     |                       |
     |                       required by a bound introduced by this call
     |
     = help: the trait `Clone` is implemented for `[T; N]`
note: required by a bound in `Lrc::<T>::make_mut`
    --> /home/njn/dev/rust1/library/alloc/src/rc.rs:1191:9
     |
1191 | impl<T: Clone> Rc<T> {
     |         ^^^^^ required by this bound in `Rc::<T>::make_mut`

error[E0599]: no method named `push` found for mutable reference `&mut [NamedMatch]` in the current scope
   --> compiler/rustc_expand/src/mbe/macro_parser.rs:280:25
    |
280 |                 matches.push(m);
    |                         ^^^^ method not found in `&mut [NamedMatch]`

error[E0277]: the trait bound `[NamedMatch]: Clone` is not satisfied
    --> compiler/rustc_expand/src/mbe/macro_parser.rs:591:35
     |
591  |                     Lrc::make_mut(&mut eof_mp.matches);
     |                     ------------- ^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `[NamedMatch]`
     |                     |
     |                     required by a bound introduced by this call
     |
     = help: the trait `Clone` is implemented for `[T; N]`
note: required by a bound in `Lrc::<T>::make_mut`
    --> /home/njn/dev/rust1/library/alloc/src/rc.rs:1191:9
     |
1191 | impl<T: Clone> Rc<T> {
     |         ^^^^^ required by this bound in `Rc::<T>::make_mut`

error[E0277]: the size for values of type `[NamedMatch]` cannot be known at compilation time
   --> compiler/rustc_expand/src/mbe/macro_parser.rs:592:51
    |
592 |                     let matches = Lrc::try_unwrap(eof_mp.matches).unwrap().into_iter();
    |                                   --------------- ^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |                                   |
    |                                   required by a bound introduced by this call
    |
    = help: the trait `Sized` is not implemented for `[NamedMatch]`
note: required by a bound in `Lrc::<T>::try_unwrap`
   --> /home/njn/dev/rust1/library/alloc/src/rc.rs:360:6
    |
360 | impl<T> Rc<T> {
    |      ^ required by this bound in `Rc::<T>::try_unwrap`

error[E0599]: the method `unwrap` exists for enum `Result<[NamedMatch], Rc<[NamedMatch]>>`, but its trait bounds were not satisfied
   --> compiler/rustc_expand/src/mbe/macro_parser.rs:592:67
    |
592 |                     let matches = Lrc::try_unwrap(eof_mp.matches).unwrap().into_iter();
    |                                                                   ^^^^^^ method cannot be called on `Result<[NamedMatch], Rc<[NamedMatch]>>` due to unsatisfied trait bounds
    |
    = note: the following trait bounds were not satisfied:
            `[NamedMatch]: Sized`

error[E0277]: the size for values of type `[NamedMatch]` cannot be known at compilation time
   --> compiler/rustc_expand/src/mbe/macro_parser.rs:592:35
    |
592 |                     let matches = Lrc::try_unwrap(eof_mp.matches).unwrap().into_iter();
    |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `[NamedMatch]`
note: required by a bound in `Result`
   --> /home/njn/dev/rust1/library/core/src/result.rs:502:17
    |
502 | pub enum Result<T, E> {
    |                 ^ required by this bound in `Result`

error[E0308]: mismatched types
   --> compiler/rustc_expand/src/mbe/macro_parser.rs:625:57
    |
625 |         self.cur_mps.push(MatcherPos { idx: 0, matches: self.empty_matches.clone() });
    |                                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Rc<[NamedMatch]>`, found `Rc<Vec<NamedMatch>>`
    |
    = note: expected struct `Lrc<[NamedMatch]>`
               found struct `Lrc<Vec<NamedMatch>>`
```

Reply:
Ugh. `Clone` requiring `Sized` was a mistake :(

Reply:
Perhaps, but that's not the only problem here.